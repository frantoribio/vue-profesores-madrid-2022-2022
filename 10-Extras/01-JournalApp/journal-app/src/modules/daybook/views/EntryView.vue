<template>
  <!-- Transicciones me pidde un solo div!!! -->
  <transition
    name="fade"
    mode="out-in"
  >
    <div v-if="myEntry">
      <div class="flex h-20 flex-row items-center justify-between p-2">
        <div>
          <span class="text-lg font-bold text-primary">{{ entryDate.day }}</span>
          <span class="mx-2 text-lg font-semibold text-base-content">{{ entryDate.month }}</span>
          <span class="mx-1 text-lg font-light text-base-content">{{ entryDate.year }}, </span>
          <span class="mx-1 text-lg font-light text-base-content">{{ entryDate.dayOfWeek }}.</span>
        </div>
        <div>
          <input
            v-show="false"
            type="file"
            @change="selectedPicture($event)"
            accept="image/*"
            ref="fileInput"
          />
          <button
            data-testid="entryview-deletebutton"
            v-if="props.id !== 'new'"
            class="btn btn-error mx-2 gap-2"
            @click="deleteEntry()"
          >
            <Icon
              icon="fa:trash"
              :inline="true"
              class="h-5 w-5"
            />
            Borrar
          </button>
          <button
            class="btn btn-primary mx-2 gap-2"
            @click="onSeletedPicture()"
          >
            <Icon
              icon="fa:upload"
              :inline="true"
              class="h-5 w-5"
            />
            Subir foto
          </button>
        </div>
      </div>
      <hr class="border-t border-primary" />
      <div class="flex h-full flex-col px-3">
        <textarea
          data-testid="entryview-textinput"
          v-model.trim="textInput"
          placeholder="¿Qué sucedió hoy?"
          class="ext-base-content textarea textarea-primary my-2 h-full w-full rounded-md bg-base-100"
        ></textarea>
      </div>
      <img
        v-if="imgUrl"
        :src="imgUrl"
        alt="imagen"
        class="fixed right-10 bottom-32 m-2 w-48 rounded-lg border-4 border-base-200 object-cover shadow-xl"
      />
      <FavButton
        data-testid="entryview-savebutton"
        icon="fa-regular:save"
        class="fixed right-10 bottom-14 h-12 w-12 shadow-md"
        @click="saveEntry()"
      />
    </div>
  </transition>
</template>

<script setup>
  import FavButton from '@/components/FabButton.vue'
  import getDateParsered from '@/helpers/dateparser.js'
  import UserStore from '@/stores/users'
  import { Icon } from '@iconify/vue'
  import Swal from 'sweetalert2'
  import { computed, ref, watch } from 'vue'
  import { useRouter } from 'vue-router'
  import JournalStore from '../stores/journal'

  const props = defineProps({
    id: {
      type: String,
      required: true,
    },
  })

  const router = useRouter()
  const journalStore = JournalStore()
  const userStore = UserStore()

  const myEntry = ref(null) // referencia a mi entrada
  const textInput = ref(null) // referencia al texto de entrada
  const imgUrl = ref(null) // referencia al url de la imagen
  const fileImage = ref(null) // referencia al dichero de la imagen
  const fileInput = ref(null) // referencia al input file

  const loadNewEntry = () => {
    myEntry.value = {
      text: '',
      picture: null,
      date: new Date().toDateString(),
    }
  }

  const checkEntry = () => {
    myEntry.value = journalStore.getEntryById(props.id)
    if (!myEntry.value) {
      // console.log('no existe la entrada')
      router.push({ name: 'daybook-no-entry' })
    }
  }

  const initData = () => {
    textInput.value = myEntry.value.text // el texto siempre será el de la entrada
    imgUrl.value = myEntry.value.picture
    fileImage.value = null
  }

  const loadEntry = () => {
    // Si es new
    if (props.id === 'new') loadNewEntry()
    // Comprobamos que existe
    else checkEntry()
    // Iniciamos los datos
    initData()
  }

  loadEntry()

  const saveNewEntry = async () => {
    // console.log('createEntry')
    let id = String(Date.now()) // Cuidado que todo son string, para luego firebase!!!
    await journalStore.createEntry(
      {
        id: id,
        text: textInput.value,
        picture: myEntry.value.picture,
        date: myEntry.value.date,
        uid: userStore.user.uid,
      },
      fileImage.value
    )
    router.push({ name: 'daybook-entry', params: { id: id } })
  }

  const updateEntry = async () => {
    //console.log('updateEntry')
    await journalStore.updateEntry(
      {
        id: myEntry.value.id,
        text: textInput.value,
        date: myEntry.value.date, // Podriamos poner new Date y obrendríamos la fecha de la actualizacion
        picture: myEntry.value.picture,
        uid: userStore.user.uid,
      },
      fileImage.value
    )
  }

  const saveEntry = async () => {
    // No podemos dejar el campo vacío
    if (textInput.value.trim() === '') {
      Swal.fire({
        title: '¡No puedes dejar el campo vacío!',
        text: 'Por favor, escribe algo en el campo de texto.',
        icon: 'error',
        confirmButtonText: 'Entendido',
      })
      return
    }

    // Comenzamos
    new Swal({
      title: 'Espere por favor',
      allowOutsideClick: false,
    })
    Swal.showLoading()
    if (props.id === 'new') {
      saveNewEntry()
    } else {
      updateEntry()
    }
    Swal.fire('¡Listo!', 'La entrada se ha guardado correctamente', 'success')
  }

  const deleteEntry = async () => {
    const { isConfirmed } = await Swal.fire({
      title: '¿Estás seguro?',
      text: '¡No podrás recuperar la entrada una vez borrada!',
      icon: 'warning',
      showDenyButton: true,
      confirmButtonText: 'Sí, borrar',
    })
    if (isConfirmed) {
      Swal.fire({
        title: 'Espere por favor',
        allowOutsideClick: false,
      })
      Swal.showLoading()
      await journalStore.deleteEntry(myEntry.value)
      router.push({ name: 'daybook-no-entry' })
      Swal.fire('¡Listo!', 'La entrada se ha eliminado correctamente', 'success')
    }
  }

  const selectedPicture = (event) => {
    fileImage.value = event.target.files[0]
    if (!fileImage.value) {
      imgUrl.value = null
      fileImage.value = null
      return
    }
    const reader = new FileReader()
    reader.onload = () => (imgUrl.value = reader.result)
    reader.readAsDataURL(fileImage.value)
  }

  const onSeletedPicture = () => fileInput.value.click()

  // La fecha es computada
  const entryDate = computed(() => {
    return getDateParsered(myEntry.value.date)
  })

  // Compruebo que existe la entrada añado un watcher!!
  watch(
    () => props.id,
    () => {
      loadEntry()
    }
  )

  // Observamos el id y sus cambios
</script>

<style lang="scss" scoped>
  .fade-enter-active,
  .fade-leave-active {
    transition: opacity 0.5s ease;
  }
  .fade-enter-from,
  .fade-leave-to {
    opacity: 0;
  }
</style>
