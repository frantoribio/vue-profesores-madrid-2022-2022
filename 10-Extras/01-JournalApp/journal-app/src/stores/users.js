import users from '@/data/users'
import espera from '@/helpers/espera'
import JournalStore from '@/modules/daybook/stores/journal'
import { defineStore } from 'pinia'

export const UserStore = defineStore({
  id: 'UserStore',
  state: () => ({
    user: null,
  }),

  getters: {
    isLoggedIn: (state) => state.user !== null,
  },

  actions: {
    setUser(user) {
      this.user = user
    },

    async logIn() {
      await espera(500)
      // TODO: Compruebas la sesi√≥n en remoto o Firebase
      this.setUser(users[0])
      // Le cargamos las entradas del usuario!!!
      const journalStore = JournalStore()
      journalStore.fetchEntries()
    },

    async logOut() {
      await espera(1000) // Solo por perder el tiempo y parezca asincrono
      // Le limpiamos los datos del usuario!!!
      const journalStore = JournalStore()
      journalStore.clearData()
      this.setUser(null)
    },
  },
})

export default UserStore
