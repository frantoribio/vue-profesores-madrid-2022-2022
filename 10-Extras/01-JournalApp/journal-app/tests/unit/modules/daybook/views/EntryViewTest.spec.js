import { beforeEach, describe, expect, test, vi } from 'vitest'

import { createTestingPinia } from '@pinia/testing'
import { shallowMount } from '@vue/test-utils'
import { createRouterMock, injectRouterMock } from 'vue-router-mock'

import EntryView from '@/modules/daybook/views/EntryView.vue'

import Swal from 'sweetalert2'

// Para esperar las promesas pendientes tambien puedes usar: https://www.npmjs.com/package/flush-promises
import { setImmediate } from 'timers'
const flushPromises = () => new Promise(setImmediate)

import entries from '../stores/mocks/entries.mocks'
import users from '../stores/mocks/users.mocks'

describe('Daybook Componente -> EntryView con New', () => {
  let wrapper
  const router = createRouterMock()

  beforeEach(() => {
    vi.restoreAllMocks()
    injectRouterMock(router)
    wrapper = shallowMount(EntryView, {
      props: {
        id: 'new',
      },
      global: {
        plugins: [
          createTestingPinia({
            initialState: {
              JournalStore: { entries: entries },
              UserStore: { user: users[0] },
            },
          }),
        ],
      },
    })
  })

  test('Debe hacer match con el snapshot', () => {
    expect(wrapper.text()).toMatchSnapshot()
  })

  test('El campo de texto debe estar vacío', () => {
    const textinput = wrapper.find('[data-testid="entryview-textinput"]')
    expect(textinput.element.value).toBe('')
  })

  test('El boton borrar no debe estar visible', () => {
    const deleteButton = wrapper.find('[data-testid="entryview-deletebutton"]')
    expect(deleteButton.exists()).toBe(false)
  })

  test('Debe crear una entrada nueva y redirigirnos a ella', async () => {
    const textinput = wrapper.find('[data-testid="entryview-textinput"]')
    textinput.setValue('text-new')
    await textinput.trigger('input')
    const saveButton = wrapper.find('[data-testid="entryview-savebutton"]')
    // pulsamos click
    await saveButton.trigger('click')
    expect(wrapper.vm.journalStore.createEntry).toHaveBeenCalled()
    // Nos debe llevar a esta entrada
    expect(router.push).toHaveBeenCalledWith({
      name: 'daybook-entry',
      params: { id: expect.any(String) },
    })
  })
})

describe('Daybook Componente -> EntryView con id', () => {
  let wrapper
  const router = createRouterMock()

  beforeEach(() => {
    vi.restoreAllMocks()
    injectRouterMock(router)
    wrapper = shallowMount(EntryView, {
      props: {
        id: entries[0].id,
      },
      global: {
        plugins: [
          createTestingPinia({
            initialState: {
              JournalStore: { entries: entries },
              UserStore: { user: users[0] },
            },
          }),
        ],
      },
    })
  })

  test('Debe hacer match con el snapshot', () => {
    expect(wrapper.text()).toMatchSnapshot()
  })

  test('Debe debe cargar la entrada si existe', () => {
    expect(wrapper.vm.myEntry).toEqual(entries[0])
  })

  test('El campo de texto debe coicidir con la entrada', () => {
    const textinput = wrapper.find('[data-testid="entryview-textinput"]')
    expect(textinput.element.value).toBe(entries[0].text)
  })

  test('El boton borrar si debe estar visible', () => {
    const deleteButton = wrapper.find('[data-testid="entryview-deletebutton"]')
    expect(deleteButton.exists()).toBe(true)
  })

  test('Debe borrar la entrada si pulsamos el boton borrar', async () => {
    const deleteButton = wrapper.find('[data-testid="entryview-deletebutton"]')
    expect(deleteButton.exists()).toBe(true)
    // Mockeamos el sweetalert
    Swal.fire = vi.fn().mockImplementation(() => Promise.resolve({ isConfirmed: true }))
    await deleteButton.trigger('click')
    // Nos debe llevar a esta entrada
    expect(wrapper.vm.journalStore.deleteEntry).toHaveBeenCalledWith(entries[0])
    await flushPromises()
    expect(router.push).toHaveBeenCalledWith({ name: 'daybook-no-entry' })
  })

  test('Debe actualizar la entrada', async () => {
    const textinput = wrapper.find('[data-testid="entryview-textinput"]')
    textinput.setValue('text-updated')
    await textinput.trigger('input')
    const saveButton = wrapper.find('[data-testid="entryview-savebutton"]')
    // pulsamos click
    await saveButton.trigger('click')
    expect(wrapper.vm.journalStore.updateEntry).toHaveBeenCalled()
    // await flushPromises()
  })
})

describe('Daybook Componente -> EntryView con id no existe', () => {
  const router = createRouterMock()

  beforeEach(() => {
    vi.restoreAllMocks()
    injectRouterMock(router)
  })

  test('No debe existir la entrada', () => {
    let wrapper
    try {
      wrapper = shallowMount(EntryView, {
        // De esta manera mockeamos el onMounted
        props: {
          id: 'caca',
        },
        global: {
          plugins: [
            createTestingPinia({
              initialState: {
                JournalStore: { entries: entries },
                UserStore: { user: users[0] },
              },
            }),
          ],
        },
      })
    } catch (e) {
      // Si no llega aquí es que no se ha renderizado, por lo que está bien!!!
      expect(wrapper).toBeUndefined()
    }
  })
})
