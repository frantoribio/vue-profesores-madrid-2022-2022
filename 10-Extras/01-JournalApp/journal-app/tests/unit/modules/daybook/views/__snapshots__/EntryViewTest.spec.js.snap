// Vitest Snapshot v1

exports[`Daybook Componente -> EntryView con New > Debe hacer match con el snapshot 1`] = `"19Octubre2022, Miércoles. Subir foto"`;

exports[`Daybook Componente -> EntryView con id > Debe hacer match con el snapshot 1`] = `"19Octubre2022, Miércoles. Borrar  Subir foto"`;
